# Student note: My answers are all displayed at bottom of notebook and in the dictionary STUDENT_ANSWERS

This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
#API_KEY = '' #API_KEY is private. Notebook computations can be reproduced with anyone's API_KEY

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

In [2]:
#!pip install requests

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [3]:
# First, import the relevant modules
import requests


In [4]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
DATASET = 'FSE'
TICKER = 'AFX_X'
LIMIT = 'limit=1'
COLLAPSE = 'collapse=day'
Quandl_API_string = "https://www.quandl.com/api/v3/datasets/"+DATASET+"/"+TICKER+"/?"+COLLAPSE+"&"+LIMIT+"&api_key="+API_KEY
res = requests.get(Quandl_API_string)
json_data = res.json()
print(res) #2XX response code means success. 4XX response code means failure.
print('\n')
print(json_data)
print('\n')
print(json_data.keys())

<Response [200]>


{'dataset': {'id': 10095370, 'dataset_code': 'AFX_X', 'database_code': 'FSE', 'name': 'Carl Zeiss Meditec (AFX_X)', 'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704', 'refreshed_at': '2020-12-01T14:48:09.907Z', 'newest_available_date': '2020-12-01', 'oldest_available_date': '2000-06-07', 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'frequency': 'daily', 'type': 'Time Series', 'premium': False, 'limit': 1, 'transform': None, 'column_index': None, 'start_date': '2000-06-07', 'end_date': '2020-12-01', 'data': [['2020-12-01', 112.2, 112.2, 111.5, 112.0, None, 51.0, 5703.0, None, None, None]], 'collapse': None, 'order': None, 'database_id': 6129}}


dict_keys(['dataset'])


In [5]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
print(json_data.keys())
for key in json_data.keys():
    print('\nTop-level key:', key)
    print(json_data[key].keys())

dict_keys(['dataset'])

Top-level key: dataset
dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [6]:
try:  # try call is to prevent deletion of dictionary if it has already been constructed
    print('Dictionary already constructed!\n',STUDENT_ANSWERS)
except: STUDENT_ANSWERS = dict() # Reader can directly access answers to questions 3-7 with this dictionary.

In [7]:
# Dataset features of interest: Open, Low, High, Close, Traded Volume
# Checking these out from the 1 imported day from above:
for col, datum in zip(json_data['dataset']['column_names'], json_data['dataset']['data'][0]):
    print(col+" : "+ str(datum))

Date : 2020-12-01
Open : 112.2
High : 112.2
Low : 111.5
Close : 112.0
Change : None
Traded Volume : 51.0
Turnover : 5703.0
Last Price of the Day : None
Daily Traded Units : None
Daily Turnover : None


#### The 5 necessary features all have real values. Good!

# Task 1) Collect FSE AFX_X data for whole year 2017

In [8]:
# Parse root URL of data set: DATA_URL
DATA_URL = Quandl_API_string.split('?')[0]
DATA_URL

'https://www.quandl.com/api/v3/datasets/FSE/AFX_X/'

In [9]:
# Add parameters to the string: URL
URL = DATA_URL + '/?collapse=day' + '&start_date=2017-01-01' + '&end_date=2017-12-31' + '&api_key='+API_KEY
URL

'https://www.quandl.com/api/v3/datasets/FSE/AFX_X//?collapse=day&start_date=2017-01-01&end_date=2017-12-31&api_key=wBspijstPsj1rCRamZPH'

In [10]:
# Collect data for whole year 2017: r
r = requests.get(URL)
print(r)

<Response [200]>


# Task 2) Convert the returned JSON object to a python dictionary

In [11]:
# Extract initial dictionary: json_daily_2017
json_daily_2017 = r.json()

# Make sure this is the same structure as above, with one top-level key containing many other keys
print(json_daily_2017.keys())
for key in json_daily_2017.keys():
    print('\nTop-level key', key, 'contains lower-level keys:\n')
    print(json_daily_2017[key].keys())

dict_keys(['dataset'])

Top-level key dataset contains lower-level keys:

dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])


In [12]:
# Nothing is lost by "subsetting" the lone top-level key of 'dataset': data_2017
data_2017 = json_daily_2017['dataset']
print(data_2017.keys())

dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])


In [13]:
#Verify data structure is list of lists with outer list being day and inner list being columns
print(len(data_2017['data']), '\n') # Looking for a length of around 250 --> Number of trading days in the year.

#Verify that each day's entry looks similar to the single day example at start of the notebook
column_names = data_2017['column_names']
print(column_names)
print(dict(keys=column_names, values=data_2017['data'][-1]),'\n')
print(dict(keys=column_names, values=data_2017['data'][0]),'\n')

255 

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']
{'keys': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'values': ['2017-01-02', 34.99, 35.94, 34.99, 35.8, None, 44700.0, 1590561.0, None, None, None]} 

{'keys': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'values': ['2017-12-29', 51.76, 51.94, 51.45, 51.76, None, 34640.0, 1792304.0, None, None, None]} 



In [14]:
# I would prefer to have dictionary keys be column names rather than dates.
# namedtuple objects can help with this.
from collections import namedtuple

col_list = []
for col in column_names:
    col_list.append(col.replace(' ', '_')) # namedtuple apparently doesn't like spaces
print(col_list)

DayEntries = namedtuple('DayEntries', col_list)
print(DayEntries._fields)

['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded_Volume', 'Turnover', 'Last_Price_of_the_Day', 'Daily_Traded_Units', 'Daily_Turnover']
('Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded_Volume', 'Turnover', 'Last_Price_of_the_Day', 'Daily_Traded_Units', 'Daily_Turnover')


In [15]:
#Enter data rows in namedtuple form into list
list_data_by_day = []
for data_by_day in data_2017['data'][-1::-1]: #Reverse order so first entry is first trading day of year.
    list_data_by_day.append(DayEntries(*data_by_day))
print(list_data_by_day[0])
print(list_data_by_day[-1])

DayEntries(Date='2017-01-02', Open=34.99, High=35.94, Low=34.99, Close=35.8, Change=None, Traded_Volume=44700.0, Turnover=1590561.0, Last_Price_of_the_Day=None, Daily_Traded_Units=None, Daily_Turnover=None)
DayEntries(Date='2017-12-29', Open=51.76, High=51.94, Low=51.45, Close=51.76, Change=None, Traded_Volume=34640.0, Turnover=1792304.0, Last_Price_of_the_Day=None, Daily_Traded_Units=None, Daily_Turnover=None)


In [16]:
list_data_by_day[0]._asdict() # namedtuple can be expressed as dictionary for easier access

{'Date': '2017-01-02',
 'Open': 34.99,
 'High': 35.94,
 'Low': 34.99,
 'Close': 35.8,
 'Change': None,
 'Traded_Volume': 44700.0,
 'Turnover': 1590561.0,
 'Last_Price_of_the_Day': None,
 'Daily_Traded_Units': None,
 'Daily_Turnover': None}

In [17]:
# Convert to dictionary form with column names as keys and value sequences as list
final_dict = {col : [] for col in col_list}
print(final_dict.keys(),'\n')
for col in final_dict.keys():
    for data_by_day in list_data_by_day:
        details = data_by_day._asdict()
        final_dict[col].append(details[col]) # Construct dictionary
for key in final_dict.keys():
    print(key, ":", final_dict[key][:5]) # Show examples from dictionary

dict_keys(['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded_Volume', 'Turnover', 'Last_Price_of_the_Day', 'Daily_Traded_Units', 'Daily_Turnover']) 

Date : ['2017-01-02', '2017-01-03', '2017-01-04', '2017-01-05', '2017-01-06']
Open : [34.99, 35.9, 35.48, 35.02, 34.91]
High : [35.94, 35.93, 35.51, 35.2, 35.21]
Low : [34.99, 35.34, 34.75, 34.73, 34.91]
Close : [35.8, 35.48, 35.19, 35.06, 35.04]
Change : [None, None, None, None, None]
Traded_Volume : [44700.0, 70618.0, 54408.0, 48412.0, 27507.0]
Turnover : [1590561.0, 2515473.0, 1906810.0, 1692326.0, 964046.0]
Last_Price_of_the_Day : [None, None, None, None, None]
Daily_Traded_Units : [None, None, None, None, None]
Daily_Turnover : [None, None, None, None, None]


# Task 3) Calculate highest and lowest opening prices for 2017

In [18]:
#Make sure I can access Open list
# max(final_dict['Open']) gives an error due to NoneType entries
max(final_dict['Open'][:5]) # but it works when limited to a few values that I know are numeric.

35.9

In [19]:
#Find NoneType entries
Open_null_record = []
for i, entry in enumerate(final_dict['Open']):
    if type(entry) != float:
        print(type(entry), 'Index #:', i, ', Date:', final_dict['Date'][i])
        print(list_data_by_day[i], '\n')
        Open_null_record.append(i)

<class 'NoneType'> Index #: 74 , Date: 2017-04-14
DayEntries(Date='2017-04-14', Open=None, High=42.48, Low=41.985, Close=42.2, Change=None, Traded_Volume=88416.0, Turnover=3734717.0, Last_Price_of_the_Day=None, Daily_Traded_Units=None, Daily_Turnover=None) 

<class 'NoneType'> Index #: 75 , Date: 2017-04-17
DayEntries(Date='2017-04-17', Open=None, High=42.48, Low=41.985, Close=42.2, Change=None, Traded_Volume=88416.0, Turnover=3734717.0, Last_Price_of_the_Day=None, Daily_Traded_Units=None, Daily_Turnover=None) 

<class 'NoneType'> Index #: 85 , Date: 2017-05-01
DayEntries(Date='2017-05-01', Open=None, High=42.245, Low=41.655, Close=41.72, Change=-0.44, Traded_Volume=86348.0, Turnover=3606589.0, Last_Price_of_the_Day=None, Daily_Traded_Units=None, Daily_Turnover=None) 



##### Given there are only 3 values like this, I suspect it is a problem with the database rather than my code.

In [20]:
# Obtain max and min while skipping missing values
HIGHEST_OPEN = max([x for i, x in enumerate(final_dict['Open']) if i not in Open_null_record])
print('Highest open:', HIGHEST_OPEN)
LOWEST_OPEN = min([x for i, x in enumerate(final_dict['Open']) if i not in Open_null_record])
print('Lowest oepn:', LOWEST_OPEN)

Highest open: 53.11
Lowest oepn: 34.0


In [21]:
STUDENT_ANSWERS['Task3'] = {'HIGHEST_OPEN':HIGHEST_OPEN, 'LOWEST_OPEN':LOWEST_OPEN}

# Task 4) Largest change in any one day 
### (High minus Low)

In [22]:
#Find NoneType entries
High_null_record = []
for i, entry in enumerate(final_dict['High']):
    if type(entry) != float:
        print(type(entry), 'Index #:', i, ', Date:', final_dict['Date'][i])
        print(list_data_by_day[i], '\n')
        High_null_record.append(i)
if len(High_null_record) == 0:
    print('All values of \'High\' are floats!')

All values of 'High' are floats!


In [23]:
#Find NoneType entries
Low_null_record = []
for i, entry in enumerate(final_dict['Low']):
    if type(entry) != float:
        print(type(entry), 'Index #:', i, ', Date:', final_dict['Date'][i])
        print(list_data_by_day[i], '\n')
        Low_null_record.append(i)
if len(Low_null_record) == 0:
    print('All values of \'Low\' are floats!')

All values of 'Low' are floats!


In [24]:
final_dict['High_minus_Low'] = []
for i in range(len(list_data_by_day)):
    diff = final_dict['High'][i] - final_dict['Low'][i]
    final_dict['High_minus_Low'].append(diff)
print(final_dict['High'][:3], '\n')
print(final_dict['Low'][:3], '\n')
print(final_dict['High_minus_Low'][:3], '\n')

[35.94, 35.93, 35.51] 

[34.99, 35.34, 34.75] 

[0.9499999999999957, 0.5899999999999963, 0.759999999999998] 



In [25]:
min(final_dict['High_minus_Low']) # Just making sure the highs are always higher than lows.

0.18999999999999773

In [26]:
LARGEST_CHANGE = max(final_dict['High_minus_Low'])
print(LARGEST_CHANGE)

2.8100000000000023


In [27]:
STUDENT_ANSWERS['Task4'] = {'LARGEST_CHANGE': LARGEST_CHANGE}

# Task 5) What was the largest change between any two days (based on Closing Price)? 
### (I am assuming this means between any two subsequent days)

In [28]:
len(final_dict['Close'])

255

In [29]:
#Find NoneType entries
Close_null_record = []
for i, entry in enumerate(final_dict['Close']):
    if type(entry) != float:
        print(type(entry), 'Index #:', i, ', Date:', final_dict['Date'][i])
        print(list_data_by_day[i], '\n')
        Close_null_record.append(i)
if len(Close_null_record) == 0:
    print('All values of \'Close\' are floats!')

All values of 'Close' are floats!


In [30]:
close_between_days = [final_dict['Close'][i+1] - final_dict['Close'][i] for i in range(len(final_dict['Close']) - 1)]

In [31]:
low_end = min(close_between_days)
print(low_end)
high_end = max(close_between_days)
print(high_end)

-2.559999999999995
1.7199999999999989


In [32]:
LARGEST_CHANGE_BETWEEN_DAYS = low_end
STUDENT_ANSWERS['Task5'] = {'LARGEST_CHANGE_BETWEEN_DAYS':LARGEST_CHANGE_BETWEEN_DAYS}

In [33]:
print(STUDENT_ANSWERS)

{'Task3': {'HIGHEST_OPEN': 53.11, 'LOWEST_OPEN': 34.0}, 'Task4': {'LARGEST_CHANGE': 2.8100000000000023}, 'Task5': {'LARGEST_CHANGE_BETWEEN_DAYS': -2.559999999999995}}


# Task 6) What was the average daily trading volume during this year?

In [34]:
#Find NoneType entries
Traded_Volume_null_record = []
for i, entry in enumerate(final_dict['Traded_Volume']):
    if type(entry) != float:
        print(type(entry), 'Index #:', i, ', Date:', final_dict['Date'][i])
        print(list_data_by_day[i], '\n')
        Traded_Volume_null_record.append(i)
if len(Traded_Volume_null_record) == 0:
    print('All values of \'Traded_Volume\' are floats!')

All values of 'Traded_Volume' are floats!


In [35]:
MEAN_VOLUME = sum(final_dict['Traded_Volume']) / len(final_dict['Traded_Volume'])
MEAN_VOLUME

89124.33725490196

In [36]:
STUDENT_ANSWERS['Task6'] = {'MEAN_VOLUME': MEAN_VOLUME}
STUDENT_ANSWERS

{'Task3': {'HIGHEST_OPEN': 53.11, 'LOWEST_OPEN': 34.0},
 'Task4': {'LARGEST_CHANGE': 2.8100000000000023},
 'Task5': {'LARGEST_CHANGE_BETWEEN_DAYS': -2.559999999999995},
 'Task6': {'MEAN_VOLUME': 89124.33725490196}}

# Task 7) (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [37]:
def sort_numeric(input_list_array_like):
    """
    Sort a numeric list.
    
    Input:
    input_list_array_like should be type list with only numeric dtypes and odd length
    
    Output:
    median is the median value
    median_index is index value of median within list
    """
    for i, value_i in enumerate(input_list_array_like):
        count_below_or_even = 0
        for value_k in input_list_array_like:
            if value_i >= value_k:
                count_below_or_even += 1
        if count_below_or_even == 128:
            median_index = i
            median = value_i
            print(count_below_or_even, median_index, median)
    return(median, median_index)              


In [38]:
MEDIAN_VOLUME, MEDIAN_VOLUME_INDEX = sort_numeric(final_dict['Traded_Volume'])
MEDIAN_VOLUME

128 222 76286.0


76286.0

In [39]:
print(final_dict['Date'][MEDIAN_VOLUME_INDEX])
print(final_dict['Traded_Volume'][MEDIAN_VOLUME_INDEX])
print(list_data_by_day[MEDIAN_VOLUME_INDEX])

2017-11-13
76286.0
DayEntries(Date='2017-11-13', Open=48.38, High=48.61, Low=47.76, Close=48.34, Change=None, Traded_Volume=76286.0, Turnover=3681337.0, Last_Price_of_the_Day=None, Daily_Traded_Units=None, Daily_Turnover=None)


In [40]:
STUDENT_ANSWERS['Task7'] = {'MEDIAN_VOLUME':MEDIAN_VOLUME}

In [41]:
STUDENT_ANSWERS

{'Task3': {'HIGHEST_OPEN': 53.11, 'LOWEST_OPEN': 34.0},
 'Task4': {'LARGEST_CHANGE': 2.8100000000000023},
 'Task5': {'LARGEST_CHANGE_BETWEEN_DAYS': -2.559999999999995},
 'Task6': {'MEAN_VOLUME': 89124.33725490196},
 'Task7': {'MEDIAN_VOLUME': 76286.0}}

# My answers:
### {'Task3': {'HIGHEST_OPEN': 53.11, 'LOWEST_OPEN': 34.0},
### 'Task4': {'LARGEST_CHANGE': 2.8100000000000023},
### 'Task5': {'LARGEST_CHANGE_BETWEEN_DAYS': -2.559999999999995},
### 'Task6': {'MEAN_VOLUME': 89124.33725490196},
### 'Task7': {'MEDIAN_VOLUME': 76286.0}}